### PR TITLE
Add design time support

### LINKED ISSUE #16: Enable design time support
So this can be used in Scaffolding and with EF Core Power Tools.

Requires implementation of design time services/extensions, like the Npgsql extension:

https://github.com/npgsql/efcore.pg/blob/main/src/EFCore.PG.NodaTime/Scaffolding/Internal/NpgsqlNodaTimeCodeGeneratorPlugin.cs

https://github.com/npgsql/efcore.pg/blob/main/src/EFCore.PG.NodaTime/Design/Internal/NpgsqlNodaTimeDesignTimeServices.cs

https://github.com/npgsql/efcore.pg/blob/main/src/EFCore.PG.NodaTime/build/netstandard2.0/Npgsql.EntityFrameworkCore.PostgreSQL.NodaTime.targets

### PR DESCRIPTION
**Pull request by StevenRasmussen**
Fixes #16

### CODE REVIEW

**Comment by ErikEJ** [comment]
I think you need to update the .csproj as well: https://github.com/npgsql/efcore.pg/blob/main/src/EFCore.PG.NodaTime/EFCore.PG.NodaTime.csproj#L23

**Comment by StevenRasmussen** [comment]
@ErikEJ - Just updated the project as well. Thanks for taking a look at this!

**Comment by ErikEJ** [comment]
You are welcome, having design time support for Nodatime with SQL Server will be so sweet!

**Comment by ErikEJ** [comment]
Could not make it work... Noticed that other packages use Nodatime 3.0.1 !?

**Comment by ErikEJ** [comment]
Wonder if you can try this iso the strange targets file?

https://github.com/ErikEJ/EFCorePowerTools/blob/master/src/GUI/ErikEJ.EntityFrameworkCore.SqlServer.Dacpac/Design/SqlServerDacpacDesignTimeServices.cs#L12

**Comment by StevenRasmussen** [comment]
Hmmm.... ok.  Would you be able to put down the steps that you are going through to try and test this?  That way I can try to troubleshoot on my end.

**Comment by ErikEJ** [comment]
As I wrote, you can simply test with the command line tooling and a file based nuget feed.

**Comment by StevenRasmussen** [comment]
Hmmm... which command specifically are you invoking with the EF Core CLI?  I was under the impression that the migration commands used design-time features (see [here](https://docs.microsoft.com/en-us/ef/core/cli/dbcontext-creation?tabs=dotnet-core-cli)).  However,  I just setup a new project referencing a local NuGet feed with the package, added some `NodaTime` properties to one of the models, ran both the commands:
`dotnet ef migrations add InitialCreate`
`dotnet ef database update`
 The commands executed successfully and the DB was created with the new date columns that map to the NodaTime properties.  I'm probably not invoking the correct command that you are seeing an error with though.

**Comment by ErikEJ** [comment]
It's the scaffold command.

But I was thinking maybe the db column must be datetime2 to map to Nodatime, and I have only tested with datetime columns.

**Comment by ErikEJ** [comment]
OK, I tried with datetime2, and it finally worked - but is the type name Noda?

```
namespace ScaffoldingTester.Models
{
    public partial class BoolNullableTest
    {
        public int Id { get; set; }
        public bool TestBool { get; set; }
        public DateTime? Noda { get; set; }
    }
}
```

**Comment by StevenRasmussen** [comment]
I'll try the scaffold command to see what happens and let you know.  To answer your question, the DB type depends on the NodaTime type.  For example, the `Instant` type maps to a `datetime2` in SQL.  The `LocalDate` type maps to a `date` and a `LocalTime` maps to `time`.  Your class should use the NodaTime types... so in your example it would be:

```csharp
namespace ScaffoldingTester.Models
{
    public partial class BoolNullableTest
    {
        public int Id { get; set; }
        public bool TestBool { get; set; }
        public Instant? CreatedOn { get; set; }
        public LocalDate SomeDate { get;set; }
    }
}
```

**Comment by ErikEJ** [comment]
Sorry, it got confused, it is still not working.

**Comment by ErikEJ** [comment]
Maybe someone from the EF Core team can see what is wrong?

@bricelamb @ajvickers ?

**Comment by StevenRasmussen** [comment]
Interesting... I've never used the DB first approach and scaffolding.  I briefly scanned the documentation but didn't see how to specify using the NodaTime types.  Is it part of the scaffold command?  Is the NodaTime package supposed to create a new provider or something based on the SQL provider? Guessing here but is it something like:
```shell
dotnet ef dbcontext scaffold "Data Source=(localdb)\MSSQLLocalDB;Initial Catalog=YourDB" SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime
```

Or how else does the scaffold command know to reverse engineer the DB types to NodaTime types?

**Comment by ErikEJ** [comment]
You reference the nodatime package in your project, and that's it! Magic (when it works)

**Comment by ErikEJ** [comment]
So on the command line, just reference the standard SQL Server package name

**Comment by StevenRasmussen** [comment]
Wow! That does sound like magic :)  Hopefully Brice or Arthur can provide some insight into what exactly the package needs to provide so that the magic happens!

**Comment by StevenRasmussen** [comment]
I wonder if the following has something to do with the magic in the Npgsql provider [here](https://github.com/npgsql/efcore.pg/blob/c9dd0b8fef4e32d27a8e5a8d3a8b06c30a337b50/src/EFCore.PG.NodaTime/Extensions/NpgsqlNodaTimeDbContextOptionsBuilderExtensions.cs#L25):
```csharp
NpgsqlConnection.GlobalTypeMapper.UseNodaTime();
```

**Comment by ErikEJ** [comment]
No idea, let's ask @roji :smile:

File: src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/build/netstandard2.0/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime.targets [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  <Target Name="AddEFCoreSqlServerNodaTime"
+          BeforeTargets="CoreCompile"
+          DependsOnTargets="PrepareForBuild"
+          Condition="'$(DesignTimeBuild)' != 'True'"
+          Inputs="$(MSBuildAllProjects)"
+          Outputs="$(EFCoreSqlServerNodaTimeFile)">
+    <ItemGroup>
+      <EFCoreSqlServerNodaTimeServices Include="Microsoft.EntityFrameworkCore.Design.DesignTimeServicesReferenceAttribute">
+        <_Parameter1>SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime.Design.SqlServerNodaTimeDesignTimeServices, SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime</_Parameter1>
+        <_Parameter2>SimplerSoftware.EntityFrameworkCore.SqlServer</_Parameter2>
```
Comment (bricelam) [inline src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/build/netstandard2.0/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime.targets:36]:
```suggestion
        <_Parameter2>Microsoft.EntityFrameworkCore.SqlServer</_Parameter2>
```

(This is the *provider* the IDesignTimeServices should be used for)

File: src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    public class SqlServerNodaTimeDesignTimeServices : IDesignTimeServices
+    {
+        public void ConfigureDesignTimeServices(IServiceCollection serviceCollection)
+            => serviceCollection
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, InstantTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalDateTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalTimeTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, DurationTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalDateTimeTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, OffsetDateTimeTypeMappingSourcePlugin>()
```
Comment (bricelam) [inline src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:20]: I wonder if the default mappings for `datetime`, etc. will prevent these from being used...

File: src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/build/netstandard2.0/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime.targets (hunk 1 above)
Comment (bricelam) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/build/netstandard2.0/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime.targets:36]: Note, this is the "magic" Erik is talking about. When you install this package, it adds an assembly-level attribute to the app that the EF Core tool will search for and use to load additional design-time services.

File: src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs (hunk 2 above)
Comment (bricelam) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:20]: If so, you may need to override and replace the `SqlServerTypeMappingSource` service to avoid mapping them to DateTime and TimeSpan.

Comment (ErikEJ) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:20]: Let me try again with the correct provider name. Thanks!

Comment (ErikEJ) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:20]: Did not work! Maybe override SqlServerTypeMappingSource then?

File: src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    [UsedImplicitly]
+    public class SqlServerNodaTimeDesignTimeServices : IDesignTimeServices
+    {
+        public void ConfigureDesignTimeServices(IServiceCollection serviceCollection)
+            => serviceCollection
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, InstantTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalDateTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalTimeTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, DurationTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalDateTimeTypeMappingSourcePlugin>()
```
Comment (ErikEJ) [inline src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:19]: It seems odd that there are two plugins that map to datetime2 ? Sould it not just be the Instant mapping only?

File: src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs (hunk 2 above)
Comment (StevenRasmussen) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:20]: @bricelam - Thanks for the guidance!  I went ahead and committed your change regarding the provider name in the .targets file.  I can look into the `SqlServerTypeMappingSource` to see what needs to happen there.

File: src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs (hunk 3 above)
Comment (StevenRasmussen) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:19]: There are several NodaTime types that while distinct in their concepts in the CLR world, translate to the same MS SQL Server DB type.  As you noted, both the `Instant` type and the `LocalDateTime` type are mapped to the `datetime2` type in the DB.  This is because both NodaTime types require the storage of date and time information in the DB.  I suppose one could map to `datetime` and the other to `datetime2` for purposes of separating the 2 but that would be arbitrary and not really represent the actual difference between the 2 types in the CLR world (and making one of them less precise than the other for no reason).  There are also currently 2 types that map to the `time` SQL type: `LocalTime` and `Duration`.  You can read up more on the NodaTime types [here](https://nodatime.org/3.0.x/userguide/core-types).

That being said, I'm not entirely sure how that would affect the scaffolding process.  We would have to default to one or the other I suppose... and that's probably where the `SqlServerTypeMappingSource` comes into play.  I can look at what the Npgsql provider does and default to the same if it makes sense.  Although IIRC PostgreSQL natively supports all of the NodaTime types and so there must be a way in the DB itself to distinguish between all the different NodaTime types.  Unfortunately that's not the case with SQL Server.

File: src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs (hunk 2 above)
Comment (StevenRasmussen) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:20]: After looking at the `SqlServerTypeMappingSource` [here](https://github.com/dotnet/efcore/blob/34fa4029dc60168b5db16f7269f8f9645fedf592/src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs), I can see that the code will need to be refactored quite a bit.  When I wrote this I was trying to keep the code similar to what was already written in the EF Core codebase but it appears that during the 5.0 timeframe, the EF Core code was refactored quite a bit and consolidated all of the `IRelationalTypeMappingSourcePlugin`'s into a single class: `SqlServerTypeMappingSource` :)  The approach that I'm thinking of taking now is to create a new `NodaTimeSqlServerTypeMappingSource` class that inherits from the `SqlServerTypeMappingSource` class and consolidates all of the current NodaTime plugins into it.  Then in the `FindMapping` method, I will return a NodaTime relational mapping where appropriate, otherwise return the default `RelationTypeMapping` from the base implementation.  We'd have to then replace the default registration in the DI container with this new version. @bricelam - Does this sound like the appropriate way to go about this?

Comment (bricelam) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:20]: > Does this sound like the appropriate way to go about this?

No. This seems like a unique case where you want an extension to override the *default* mapping for a store type (e.g. `datetime`) at design time. IMHO, the SqlServerTypeMappingSource should only be overridden at design time and not runtime. Overriding it at runtime could potentially create conflicts with other extensions or apps that also try to override it. In other words, keep the type mapping plugins for correct runtime behavior when the users adds NodaTime properties to their entities, but override SqlServerTypeMappingSource at design-time to force mapping to NodaTime types instead of BCL types when reverse engineering a model.

It all feels a bit brittle to me since the extensibility is really only designed to **add** type mappings, but doing it this way feels the least brittle.

Comment (bricelam) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:20]: @roji Do you support mapping to NodaTime types during reverse engineer?

**Comment by bricelam** [comment]
I wonder if reverse-engineering to NodaTime types is more of a concern for code generation, and should be handled in templates somewhere instead of while transforming the database schema into an EF Core model...

File: src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs (hunk 2 above)
Comment (StevenRasmussen) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:20]: > This seems like a unique case where you want an extension to override the default mapping for a store type (e.g. datetime) at design time.

Agreed :)  So instead of replacing the default provider across the board, I would only replace it during the reverse engineering process.  Does that just happen here in the `ConfigureDesignTimeServices` method?  If so then I *think* the method should look something like:
```csharp
public void ConfigureDesignTimeServices(IServiceCollection serviceCollection)
{
    // Search for the current default 'SqlServerTypeMappingSource'
    var mappingSourceDescriptor = serviceCollection.FirstOrDefault(x =>
        x.ServiceType == typeof(IRelationalTypeMappingSourcePlugin)
        && x.ImplementationType == typeof(SqlServerTypeMappingSource));

    // Determine where in the collection the NodaTime plugins should be inserted
    var insertionIndex = mappingSourceDescriptor != null
        ? serviceCollection.IndexOf(mappingSourceDescriptor)
        : serviceCollection.Count - 1;

    // These are either added before the current default plugin (if it exists) or the end of the collection
    // I'm assuming the EFCore internals iterate over all of the 'IRelationalTypeMappingSourcePlugin' registrations
    // and use the first valid 'RelationalTypeMapping' that is returned when searching for a valid type mapping.
    // Since these are in the collection *before* the default plugin, these should override the default where appropriate
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(InstantTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(LocalDateTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(LocalTimeTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(DurationTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(LocalDateTimeTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(OffsetDateTimeTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(SqlServerNodaTimeCodeGeneratorPlugin), ServiceLifetime.Singleton));
}
```

Does that *seem* correct?

**Comment by StevenRasmussen** [comment]
That's a valid point... I would defer to your expertise on this.  Maybe @roji could comment on how it's currently handled by the Npgsql provider.

File: src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs (hunk 3 above)
Comment (roji) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:19]: PostgreSQL definitely does not have date/time store types which happen to perfectly correspond to the NodaTime CLR types. If you look at [NpgsqlNodaTimeTypeMappingSourcePlugin](https://github.com/npgsql/efcore.pg/blob/main/src/EFCore.PG.NodaTime/Storage/Internal/NpgsqlNodaTimeTypeMappingSourcePlugin.cs#L54), you'll see that both Instant and LocalDateTime map to PostgreSQL `timestamp without time zone`.

However, it's perfectly normal for more than one CLR type to be mapped to the same database type. When scaffolding, the type mapping source (or plugin) is asked to find a mapping given a store type only (string) and no CLR type, and must return a single mapping (and so a single CLR type). For example, while in Npgsql both Instant and LocalDateTime map to PostgreSQL `timestamp without time zone`, if you scaffold a column of that type you'll get an Instant as the default corresponding type.

File: src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs (hunk 2 above)
Comment (roji) [reply src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs:20]: @StevenRasmussen it's important to distinguish between a provider's TypeMappingSource implementation, and a plugin-provided TypeMappingSourcePlugin. As @bricelam wrote above, you should not be trying to provide a TypeMappingSource, since that would have to replace SqlServerTypeMappingSource, which is undesirable; this is true for both runtime and design-time - there shouldn't be any reason to replace SqlServerTypeMappingSource in any situation. You should ideally just be providing a single SqlServerNodaTimeTypeMappingSourcePlugin (note the Plugin at the end).

**Comment by roji** [comment]
Wrote some comments above, but tl;dr Npgsql's EF NodaTime plugin does support scaffolding via a single NpgsqlNodaTimeTypeMappingSourcePlugin (no replacing of the provider's TypeMappingSource neither in runtime nor in design-time).

From a cursory glance, the PR seems like it should work (though I'd consolidate the plugins into a single one). @ErikEJ @StevenRasmussen how have you attempted to test this? If you've tried taking a dependency on this plugin via `<ProjectReference>` from a test project, then IIRC that's not going to work; I think there's a bit of magic that makes the .targets file only work out-of-the-box when using a `<PackageReference>` with a nuget; to test with `<ProjectReference>` you need to manually import the plugin's .targets file in the test project's csproj.

> NpgsqlConnection.GlobalTypeMapper.UseNodaTime();

This adds NodaTime support at the Npgsql ADO level (on top of which the EF support is implemented) - it is Npgsql-specific and you don't need it, since SqlServer support is entirely done at the EF level.

**Comment by roji** [comment]
BTW please ping me again if you still can't make it work, and I'll try it out.

**Comment by StevenRasmussen** [comment]
@roji - Thanks for stepping in here and providing some feedback.

> From a cursory glance, the PR seems like it should work (though I'd consolidate the plugins into a single one)

Agreed.  When I started this project in the 3.x timeframe I tried to follow the same pattern that existed in the EF Core codebase which I believe had them all separated out.  It looks like during the 5.x timeframe they were consolidated into a single class.  I think that's a cleaner approach and will look at doing that.

> how have you attempted to test this?

@ErikEJ  mentioned the issue regarding using a `<ProjectReference>` and `<PackageReference>` and so I've been using a local folder based NuGet feed so that a `<PackageReference>` is used.  When I run the scaffold command to generate a DB context it executes successfully and even adds the `UseNodaTime` method as specified by the `SqlServerNodaTimeCodeGeneratorPlugin` (so I know its picking up this plugin as part of the generation) but the CLR types for the model are just `DateTime` instead of an `Instant` or a `LocalDate` per the existing DB schema (which map to a `datetime2` and `date` column respectively in the DB).

I guess it's still not clear to me how the scaffolding command should be picking up and using each of the `IRelationalTypeMappingSourcePlugin`'s defined in my `SqlServerNodaTimeDesignTimeServices.ConfigureDesignTimeServices` method in order to override the default type mappings from the `SqlServerTypeMappingSource`.  This is where I was coming up with the idea that the `SqlServerTypeMappingSource` needed to be replaced somehow so that it would know that if it encountered a `datetime2` column type that it should map to an `Instant` instead of a `DateTime` CLR type.  Anyway, I'm sure I'm still missing something somewhere so if you have a chance to take a look that would be much appreciated. Thanks!

**Comment by ErikEJ** [comment]
@bricelam Not sure why you think templates are needed, as mentioned by @roji this works well in EF Core Power Tools for both MySQL and PostgreSQL, cannot see why we cannot make the SQL Server support equally great?

I am able to use the plugin functionality from a local package feed, but experience the same issue as @StevenRasmussen - the UseNodaTime call is added to the generated DbContext, but no mapping of datetime2 columns to Instant takes place, so if you could have a look, @roji  that would be really great!

**Comment by roji** [comment]
I took a look, and the TypeMappingSourcePlugins simply don't seem to support finding a mapping based only on a store type - which is what scaffolding does. For example, here's [InstantTypeMappingSourcePlugin.FindMapping](https://github.com/StevenRasmussen/EFCore.SqlServer.NodaTime/blob/master/src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Storage/InstantTypeMappingSourcePlugin.cs):

```c#
public RelationalTypeMapping FindMapping(in RelationalTypeMappingInfo mappingInfo)
{
    var clrType = mappingInfo.ClrType ?? typeof(Instant);
    var storeTypeName = mappingInfo.StoreTypeName;

    return typeof(Instant).IsAssignableFrom(clrType)
        ? new InstantTypeMapping(storeTypeName ?? "datetime2", clrType)
        : null;
}
```

When scaffolding, mappingInfo.ClrType is always null, and so this returns null. You can take a look at [NpgsqlNodaTimeTypeMappingSourcePlugin](https://github.com/npgsql/efcore.pg/blob/main/src/EFCore.PG.NodaTime/Storage/Internal/NpgsqlNodaTimeTypeMappingSourcePlugin.cs) to see how to implement a type mapping source plugin that maps both ways (and handles all NodaTime types in a single plugin).

On an unrelated note, I noticed that the UseNodaTime extension is defined in the Microsoft.EntityFrameworkCore.SqlServer.NodaTime.Extensions namespace. We usually define such extensions under Microsoft.EntityFramework, so that no additional `using` directives are necessary. This is particularly important here, as the scaffolded DbContext has UseNodaTime, but the directive is missing.

**Comment by StevenRasmussen** [comment]
@roji - Thanks for checking and that makes sense now! I’ll work on refactoring the code to consolidate into a single plugin and at the same time address this issue. Hopefully it should “just work” after this :)  Thanks again!

**Comment by roji** [comment]
I think it will :) I can already see "UseNodaTime" appearing in the scaffolded DbContext, so it really is only the type mapping that isn't working properly. Once that's done, it should achieve magic status as above...

**Comment by StevenRasmussen** [comment]
Ok.  I just pushed several changes that I thought would do it but unfortunately it's still not working for some reason.  The logic for selecting the `RelationalTypeMapping` has been consolidated into the `SqlServerNodaTimeTypeMappingSourcePlugin.cs` file.  I checked and both the `StoreTypeMappings` and `ClrTypeMappings` collections are being populated correctly so that it *should* return a valid mapping based on only the store type so I'm not quite sure what the issue could be.  Sorry to ping you on this again @roji but would you mind taking another look?

**Comment by roji** [comment]
Bad news - there's an ordering issue in SqlServerTypeMappingSource which makes it try plugins only if a built-in mapping isn't found - I've opened https://github.com/dotnet/efcore/issues/24660 to track. As a workaround, you should be able to replace SqlServerTypeMappingSource with your own class which extends it, and simply overrides FindMapping to switch the lookup order.

**Comment by ErikEJ** [comment]
@StevenRasmussen I can confirm that this did not work! @roji So this worked for other provider how?

**Comment by roji** [comment]
@ErikEJ our messages crossed each other :)

**Comment by StevenRasmussen** [comment]
I think we're finally there!  I just pushed the changes for the suggested workaround and was able to reverse engineer a DB that produced properties that had NodaTime types!  Thanks everyone for your help!  @ErikEJ - If you want to double check that everything looks good on your end I will go ahead and merge and update the published NuGet package.

**Comment by ErikEJ** [comment]
Fantastic! 

I will have a look soon, maybe bump version to 5.1 doe to obsolete API etc?

**Comment by StevenRasmussen** [comment]
I had considered bumping to 5.1 but wanted to keep the library's Major.Minor in step with EFCore to reduce confusion when determining which package to install.

**Comment by ErikEJ** [comment]
tenor

**Comment by ErikEJ** [comment]
:shipit:

File: README.md [hunk 4]
```diff
@@ -26,12 +26,23 @@ Install-Package SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime
 
 Then call the `UseNodaTime()` method as part of your SqlServer configuration during the `UseSqlServer` method call:
 ```csharp
-using Microsoft.EntityFrameworkCore.SqlServer.NodaTime.Extensions;
+using Microsoft.EntityFrameworkCore.SqlServer;
```
Comment (ErikEJ) [inline README.md:29]: should this not simply be: using Microsoft.EntityFrameworkCore ?

Comment (StevenRasmussen) [reply README.md:29]: Doh!  Good catch. Thanks!

**Comment by ErikEJ** [comment]
LGTM

**Comment by roji** [comment]
Awesome that this is done! FYI https://github.com/dotnet/efcore/issues/24660 has also been merged on the EF Core side, so the SqlServerTypeMappingSource hack can be removed for 6.0 at some point.

**Comment by StevenRasmussen** [comment]
Thanks @roji !!
